### PR TITLE
Distributed spawn improvements

### PR DESCRIPTION
**Pull request by rickard-green**
This PR provide functionality for atomically spawn and monitor processes on remote nodes (which has been possible to do on local nodes since OTP R11B). It also improves the efficiency of all distributed spawn operations by extending the distribution protocol with new spawn primitives. The VM can by this spawn the process directly on reception of the spawn request and by this avoid having to spawn via the `net_kernel` process.

The old protocol which spawned via the `net_kernel` process caused decoding of the data (from external term format; which can be lots of work if the argument list is large) to be done by `net_kernel` as well as extra copying of data between `net_kernel` and the new process. Since `net_kernel` is an important server that has other important tasks to handle this is quite unfortunate. In order to reduce the amount of work needed to be done by the distribution controller (which is the bottleneck of a connection), decoding of the argument list is now done by the newly created process itself.

A new asynchronous [`spawn_request/5`](https://erlang.org/~rickard/OTP-15251/erts-10.5.6/doc/html/erlang.html#spawn_request-5) BIF has been introduced (which also exist in various convenience versions [`spawn_request/1-4`](https://erlang.org/~rickard/OTP-15251/erts-10.5.6/doc/html/erlang.html#spawn_request-1) with default values). This, for example, makes it possible to spawn in parallel on multiple different nodes.

The `spawn_request()` BIF is also the primitive used for all distributed spawn operations assuming the remote node supports it. If the remote node does not support it, `spawn()`, `spawn_link()`, and `spawn_opt()` will fall back to using the old protocol spawning via `net_kernel`.

A distributed [`spawn_monitor()`](https://erlang.org/~rickard/OTP-15251/erts-10.5.6/doc/html/erlang.html#spawn_monitor-4) BIF has been introduced.

Distributed [`spawn_opt()`](https://erlang.org/~rickard/OTP-15251/erts-10.5.6/doc/html/erlang.html#spawn_opt-5) now support the `monitor` option if the remote node supports the new distributed spawn protocol.

[Sequential tracing](https://erlang.org/~rickard/OTP-15251/lib/kernel-6.5/doc/html/seq_trace.html) has been updated to generate trace messages showing the data flow between parent process and child process during the spawn operation.

The new distribution control messages due to the `spawn_request()` BIF:
* [`SPAWN_REQUEST`](https://erlang.org/~rickard/OTP-15251/erts-10.5.6/doc/html/erl_dist_protocol.html#SPAWN_REQUEST)
* [`SPAWN_REQUEST_TT`](https://erlang.org/~rickard/OTP-15251/erts-10.5.6/doc/html/erl_dist_protocol.html#SPAWN_REQUEST_TT)
* [`SPAWN_REPLY`](https://erlang.org/~rickard/OTP-15251/erts-10.5.6/doc/html/erl_dist_protocol.html#SPAWN_REPLY)
* [`SPAWN_REPLY_TT`](https://erlang.org/~rickard/OTP-15251/erts-10.5.6/doc/html/erl_dist_protocol.html#SPAWN_REPLY_TT)

API changes 2019-12-10:
* Reply tag for `spawn_request()` is now an option `{reply_tag, Tag}` instead of an argument to the BIF
* An option `{timeout, Timeout}` has been added for `spawn_opt()` and `spawn_request()`

### CODE REVIEW

File: lib/kernel/doc/src/seq_trace.xml [hunk 1]
```diff
@@ -104,19 +105,13 @@ seq_trace:set_token(OldToken), % activate the trace token again
           <tag><c>set_token(send, <anno>Bool</anno>)</c></tag>
           <item>
             <p>A trace token flag (<c>true | false</c>) which
-              enables/disables tracing on message sending. Default is
-              <c>false</c>.</p>
-          </item>
-          <tag><c>set_token('spawn', <anno>Bool</anno>)</c></tag>
```
[the remark this answers is not in the document]
Comment (rickard-green) [reply lib/kernel/doc/src/seq_trace.xml]: `spawn` has never been released (has only lived in master), so no.

File: erts/preloaded/src/erlang.erl [hunk 2]
```diff
@@ -20,9 +20,12 @@
...
+	 spawn_monitor/1,spawn_monitor/2,
+	 spawn_monitor/3,spawn_monitor/4,
 	 spawn_opt/2,spawn_opt/3,spawn_opt/4,spawn_opt/5,
-	 disconnect_node/1]).
+         spawn_request/1, spawn_request/2,
+         spawn_request/3, spawn_request/4, spawn_request/5,
+         spawn_request/6, disconnect_node/1]).
 -export([spawn/1, spawn_link/1, spawn/2, spawn_link/2]).
 -export([yield/0]).
 -export([crasher/6]).
```
[the remark this answers is not in the document]
Comment (rickard-green) [reply erts/preloaded/src/erlang.erl]: Yes

File: erts/emulator/test/process_SUITE.erl [hunk 3]
```diff
@@ -2325,6 +2342,602 @@ handle_event(Event, Pid) ->
...
+spawn_request_bif(Config) when is_list(Config) ->
+    io:format("Testing spawn_request() on local node...~n"),
+    spawn_request_bif_test(node()),
+    {ok, Node} = start_node(Config),
+    io:format("Testing spawn_request() on remote node...~n"),
+    spawn_request_bif_test(Node),
+    stop_node(Node),
+    ok.
+                       
+spawn_request_bif_test(Node) ->
```
[the remark this answers is not in the document]
Comment (rickard-green) [reply erts/emulator/test/process_SUITE.erl]: Yes

File: erts/doc/src/erlang.xml [hunk 4]
```diff
@@ -6443,22 +6485,367 @@ true</pre>
...
+        <p>
+	  Asynchronously send a spawn request. Returns a request
+	  identifier <c><anno>ReqId</anno></c>.
+	</p>
+	
+	<p>
+	  If the spawn operation is successful, a new process is
+	  created on the node identified by <c><anno>Node</anno></c>
+	  and the caller will be sent a message on the form
+	  <c>{<anno>Tag</anno>, <anno>ReqId</anno>, ok, Pid}</c>
```
[the remark this answers is not in the document]
Comment (rickard-green) [reply erts/doc/src/erlang.xml]: Provide an easy to use mapping to the user. As an example, if implementing `rpc:multicall()` using `spawn_request()` one can use `{Id, Node}` where `Id` is a reference specific to a specific `multicall()` call. `Id` makes it easy to know that the reply concerns a specific muliticall and the `Node` gives you a mapping to the node that the request wes sent to also on errors. Upon reception of `{{Id, Node}, ReqId, error, notsup}` you directly know what node(s) you need perform an old style call request to.

File: lib/compiler/src/beam_ssa_recv.erl [hunk 5]
```diff
@@ -165,19 +165,21 @@ recv_opt_makes_ref([I|Is], RecvLbl, Blocks, Acc) ->
...
-    case Func of
-        {make_ref,0} ->
+    case MFA of
+        {erlang,make_ref,0} ->
             {yes,Dst};
-        {monitor,2} ->
+        {erlang,monitor,2} ->
             {yes,Dst};
-        {spawn_monitor,A} when A =:= 1; A =:= 3 ->
+        {erlang,spawn_request,A} when 1 =< A, A =< 6 ->
```
Comment (bjorng) [inline lib/compiler/src/beam_ssa_recv.erl]: I suggest that you add a test case to `lib/compiler/test/receive_SUITE` to ensure that this new clause is covered.

File: erts/doc/src/erlang.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (rickard-green) [reply erts/doc/src/erlang.xml]: Note that the `Tag` is more than a nice to have feature, but actually quite essential. In the scenario above, `rpc:multicall()` would have to explicitly match each reply one at a time. If the last reply is a `{spawn_reply, ReqId, error, notsup}` the old style call request would be delayed until all other replies had arrived, which would be very bad.

File: lib/kernel/doc/src/seq_trace.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (garazdawi) [reply lib/kernel/doc/src/seq_trace.xml]: ah, yes of course.

File: erts/doc/src/erlang.xml [hunk 6]
```diff
@@ -6250,6 +6250,23 @@ true</pre>
       </desc>
     </func>
 
+    <func>
+      <name name="spawn_monitor" arity="2" since=""/>
```
Comment (sverker) [inline erts/doc/src/erlang.xml]: Missing since="OTP @OTP-15251@"

File: erts/emulator/test/process_SUITE.erl [hunk 7]
```diff
@@ -2332,6 +2351,821 @@ handle_event(Event, Pid) ->
...
+    spawn_request_bif_test(true, node()),
+    io:format("Testing spawn_request() on local node with Node...~n"),
+    spawn_request_bif_test(false, node()),
+    {ok, Node} = start_node(Config),
+    io:format("Testing spawn_request() on remote node...~n"),
+    spawn_request_bif_test(false, Node),
+    stop_node(Node),
+    ok.
+                       
+spawn_request_bif_test(Local, Node) ->
```
Comment (garazdawi) [inline erts/emulator/test/process_SUITE.erl]: Should we add a testcase with a "large" enough argument to trigger dist message fragmentation?

**Pull request merged**